tools/db_crashtest.py

```diff
@@ -801,6 +801,24 @@ def blackbox_crash_main(args, unknown_args):
 
         time.sleep(1)  # time to stabilize before the next run
 
+    # We should run the test one more time with VerifyOnly setup and no-timeout
+    # Only do this if the tests are not failed for total-duration
+    print("Running final time for verification")
+    ops_count = max(cmd_params.get("reopen", 0), 1)
+    cmd_params.update({"ops_per_thread": ops_count})
```

Comment:
I address this by adding a flag to run post_verification_only. Currently, at the end of all runs of blackbox crash test I am running post_validation_only for 1 final time.


To me, in the future,  if we might use this flag to split the db_stress runs into 2. In each round we will run 1/pre-validation, operate (crash somewhere randomly) and 2/post-validation only. I also put a todo to report statistics about the validation like how many keys are validated by each method to reason about the work validation did.

Reply:
> I am running post_validation_only for 1 final time.
> 2/post-validation only.

Hi :) sorry I just got to catch up on this. One meta question for my own learning: the PR reads to me that since the current post_validation_only skips the pre-run validation and db operation, it is actually a pre-run validation of the final uninterrupted run. If that's the case, why don't we name and implement it as pre_validation_only? It seems to me that it could satisfy the PR purpose `without timeout, to ensure validation is executed for all keys, at least once.`

Extended from `bypass either the pre-run or post-run verification`, I could see in total there are several blackbox verification cases as below, where one/some of them might not be useful at all:
(1)pre-run verification only (my hypothetical `pre_validation_only`, which IMO essentially same as (4) but correct me if I miss anything)
(2)pre-run verification + db operation (most of the blackbox tests right now)
(3)pre-run verification + db operation + post-run verification (Not sure if such post-run is still useful)
(4)post-run verification only (essentially same as (1) IMO)
(5)db operation + post-run verification (not sure if such db operation is still legit when we don't guarantee it to start from a clean point (i.e, verified by pre-run verification)

Reply:
Extended from `bypass either the pre-run or post-run verification`, I could see in total there are several blackbox verification cases as below, where one/some of them might not be useful at all:
(1)pre-run verification only (my hypothetical `pre_validation_only`, which IMO essentially same as (4) but correct me if I miss anything)
(2)pre-run verification + db operation (most of the blackbox tests right now)
(3)pre-run verification + db operation + post-run verification (Not sure if such post-run is still useful)
(4)post-run verification only (essentially same as (1) IMO)
(5)db operation + post-run verification (not sure if such db operation is still legit when we don't guarantee it to start from a clean point (i.e, verified by pre-run verification)

Reply:
Hi @hx235. Thanks a lot for the comments and enumerating all scenarios. My understanding is the same, 1 and 4 are the same. Essentially we want a validation_only flag and do not kill during validation_only runs. I am happy to change the flag name if you think it makes more sense to the reader.

Reply:
Yes - I believe the name `validation_only` made more sense. What's more, I believe by implementing this flag essentially as a "pre-run validation only" (instead of now as a "post-run validation only") can have an extra benefit of working with `FLAGS_preserve_unverified_changes`. 
- `FLAGS_preserve_unverified_changes=true` allows us to keep the .state, .trace, SST files etc from the previous run when current run's crash-recovery process fails. It's useful for investigating such failure by inspecting the trace file/SST files some time. See https://github.com/facebook/rocksdb/commit/03c4ea26bb6797873c394f569478c79fe6fc67e8 for more

Reply:
Hi @hx235 , I change the flag name to verification_only and also addressed your comment of premature printing of Crash Consistency Verification Passed message. I will be working on running refactoring the implementation to run pre-run validation instead of post-run validation.